### PR TITLE
Error loading jupyter and spyder due to permission issue

### PR DESCRIPTION
**Pull request by francisco-dlp**
By default WinPython sets the home env variable to the "settings" folder in the WinPython folder. However, non-admin users don't have write access to this folder, which means that any program to writes to that folder fails.

This fixes the issue by providing scripts to start spyder, qtconsole and notebook without modifying the default HOME env value.
  
This also removes the matplotlibrc file that is not longer necessary since we're not going to release any py2 version of the bundle.

### CODE REVIEW

**Comment by vidartf** [comment]
There are a few other options here that could be more usable:
 1. Simply to delete the "settings" folder from the WinPython folder. If it is missing, WinPython will use the users profile folder as its settings store, no matter how it is launched. This will however lose any useful default settings distributed with WinPython (so will this PR). 
 2. Include the settings folder as normal, but if installing as an Admin, then set grant write permission to the settings folder for current user/all (http://nsis.sourceforge.net/AccessControl_plug-in).
 3. A mix of 1. and 2.: Install the contents of the settings folder to the users' profile folder as in 1., and set write permissions as in 2.

**Comment by francisco-dlp** [comment]
I have tried 1 and it didn't work for me, despite what the WinPython docs say. Even if it worked, the current implementation has the advantage that the settings folder remains in the winpython folder, and, if an user decides to move the bundle to an usb key, he may find it useful. 2 doesn't sound right as users shouldn't share the config files. 3 may work for the user that install it, but all others won't have the files copied to their profile. Another option is to copy it to all users, but that would have the same drawback as 2. We could create a script to do the right thing, but unless something is actually broken I am for leaving things as they are.

**Comment by vidartf** [comment]
Strange.. 1 works for me, though I had to restart the WinPython command prompt to get the updated environment variables.

Not entirely sure what this PR does. Doesn't it basically do the same as 1 should do, except it only works when started via the custom scripts?

**Comment by francisco-dlp** [comment]
It depends on how you define "works". The problem that I'm trying to solve is being able to run the scripts when the user has no writing permissions in the bundle installation folder. Removing the settings folder doesn't fix this issue because winpython tries to recreate the folder and if the user doesn't have admin rights it fails.

As I mentioned before, even if it worked, the current implementation has the advantage that, as we don't remove the settings folder, one could simply copy the bundle folder to an usb key and have a perfectly movable winpython distribution. Moreover, this issue was present in previous releases. Therefore, the settings folder is present in all installations of hyperspy bundle and it is not removed by the uninstaller because the uninstaller only removes the files that it has created itself. Therefore, even if we could really simply go for 1, that wouldn't work for the users that are upgrading their installation.

Last but not least, this implementation works AFAIK. If it ain't broken, don't fix ;)

**Comment by vidartf** [comment]
Not sure if I follow/agree with all of your points:

> winpython tries to recreate the folder

Then this should then be reported as an issue to WinPython independent of which solution we pick, as the docs state that it should work. I cannot reproduce that behavior here, so it should either be for a specific version of WinPython, or for special setups.

> the current implementation has the advantage that, as we don't remove the settings folder, one could simply copy the bundle folder to an usb key and have a perfectly movable winpython distribution.

You should also be able to this with a deleted settings folder, and the final behavior will be the same: The installation on the USB stick will use the settings present on whatever system it is installed on, not the settings on the USB stick. The "settings" folder should be writable when on the USB stick, but python will not use it. A possible solution to this would be to modify this PRs patching to have `env.bat` check whether the settings folder is writable, and set the environment based on the result.



> Moreover, this issue was present in previous releases. Therefore, the settings folder is present in all installations of hyperspy bundle and it is not removed by the uninstaller because the uninstaller only removes the files that it has created itself.

I think you misunderstand how the uninstaller works: When uninstalling the bundle, we always use the uninstaller created during the PREVIOUS installation (uninstaller is in the root directory of the installed bundle). The new uninstaller is only used to remove the NEW installation at some point in the future. Otherwise we would not have to bundle a replacement for the broken 0.6 uninstaller.

> Last but not least, this implementation works AFAIK.

There remains a serious issue (at least in previous version of WinPython, not tested in the newest version): When you launch the WinPython command prompt, it doesn't actually use `env.bat`. It does the same behavior as the default implementation of `env.bat` behind the scenes, but never picks up on changes in `env.bat`. If you launch python or hyperspy from the command prompt you would still have the issue that this PR proposes to fix.

**Comment by francisco-dlp** [comment]
Removing the `settings` folder is not working for me using the latest plain WinPython-64bit-3.5.1.2.exe. After investigating the issue further I'll consider reporting it to the winpython developers.

> You should also be able to this with a deleted settings folder, and the final behavior will be the same

I don't think so. If you delete the settings folder you have to recreate it to make the distribution fully portable again, otherwise it will write the setting to the user profile, which is not what you normally want when running from an usb key. Note that the current implementation doesn't patch the `env.bat` script in `Scripts`, it creates a new one. Furthermore, all the launchers in the root folder remain where they are. In other words, the WinPython distribution remains pristine, except for the few extra packages that we install and the extra `hspy_scripts` folder.

> I think you misunderstand how the uninstaller works: 

Let's see if I can understand the uninstaller that I wrote. The uninstaller removes all the files that it creates during installation. This includes the settings folder. But, if the user launches e.g. hyperspy with any previous version of the bundle, ipython creates new files in the `settings` folder. Because the installer didn't create those files, it won't delete them when uninstalling during the upgrade process, and therefore the `settings` folder will remain there. Therefore, this issue will not be fixed for those upgrading. Furthermore, users who are not upgrading can easily fall into this problem by simply executing any of the launchers in the bundler root folder, as those may recreate the settings folder.

> There remains a serious issue

I have fixed the issue in my latest commit. Please consider merging.

File: hspy_bundle/configure_installer.py [hunk 1]
```diff
@@ -182,6 +227,46 @@ def create_delete_macros(self):
...
+        This created 4 scripts:
+
+        * env.bat: Taken from WinPython and patched not to use settings as
+            home directory
+        * jupyter_qtconsole.bat: uses our env.bat and add the ability to
+            to start in the directory specified by the first parameter
+        * jupyter_notebook.bat: uses our env.bat.
+        * spyder.bat: uses our env.bat.
+
+        """
```
Comment (to266) [inline hspy_bundle/configure_installer.py]: This documentation needs updating with `python.bat` and `cmd.bat`

**Comment by vidartf** [comment]
> otherwise it will write the setting to the user profile, which is not what you normally want when running from an usb key.

Won't this PR do the same?


> Let's see if I can understand the uninstaller that I wrote.

Sorry, I didn't mean to be that confrontational. Let me break down how I understand the uninstaller to work, and then you can point out where you disagree:

 - Let's say the user has the 0.8.1 bundle previously installed (but it could be whatever), and wants to install the new, magnificent 0.8.3 bundle.
 - Upon starting the installation of 0.8.3, the installer checks the registry, and sees that a previous version is installed. It asks the user if it want to uninstall the previous version.
 - The user clicks to uninstall the previous version. The installer then finds the path to the 0.8.1 uninstaller in the registry, and launches it.
 - The 0.8.1 uninstaller removes `settings` folder, since it was installed with it, independent of what the 0.8.3 installer/uninstaller does.
 - The 0.8.3 installer starts installing the new files. Which may or may not include the `settings folder`. It then installs the new 0.8.3 uninstaller.
 - After 0.9 is released, the user installs the new bundle, again asking to uninstall the previous version. The uninstaller of 0.8.3 is run. If 0.8.3 was installed with a `settings` folder, that folder will be removed.

**Comment by vidartf** [comment]
> won't these launchers be broken no matter how we fix this?

To expand on this, lets assume root installation folder requires admin privileges to write:
 - If the `settings` folder is missing, the launcher will try to create it and fail, meaning the launcher cannot be used. However, since it cannot create the `settings` folder, it will not affect the install.
 - If the `settings` folder is present, the launchers will work (?), but anything that tries to write to a file in the folder (or create a subfolder) will fail, and therefore be broken for all intents and purposes(?).

**Comment by francisco-dlp** [comment]
@vidartf, I think that this PR fixes the issue that it reports. As you have pointed out, there is another way to fix this. Both approaches to the solution have their subtle advantages and disadvantages. For me the main advantage of this fix is that it is already implemented and tested. Perfect is the enemy of good...

**Comment by vidartf** [comment]
@francisco-dlp : My main concern is that it would break the HyperSpyUI bundle installer I made which relies on the HyperSpy bundle installer :) I already encountered this issue with it, and fixed it by moving/deleting the settings folder. I need to figure out what to do with that. My current idea is to just leave it as it is, and let that be a problem for next HyperSpyUI bundle (first official release should match HyperSpy 0.9).

**Comment by francisco-dlp** [comment]
Fixing it with this fix should be as simple as adding a new batch script to the `hspy_scripts` folder, isn't it?

**Comment by vidartf** [comment]
The way I set up the UI installer, is that you can choose when installing whether to install the HyperSpy bundle as the base or not (if not, you need to specify a valid python install). As such, it is supposed to be bundle agnostic. I'm basically trying to have a single installer file independent of your python setup, in the python spirit of having one obvious way to achieve your goal, and preferably only one. Having that installer interact well with this proposed solution would make the installer quite complex.

**Comment by francisco-dlp** [comment]
Although this is a totally separate issue, why reinventing the wheel? Once we release HyperSpy 0.9 we'll include HyperSpyUI in the bundle. Therefore, anybody wanting HyperSpyUI in the bundle will have it out of the box. Those who install using the python installers should be able to install HyperSpyUI using a standard python installer that shouldn't care about bundles, etc. Or am I missing something here?

**Comment by vidartf** [comment]
I made the bundle installer since I needed it a month ago to distribute it locally at NTNU (a distributable installer was a deliverable of my extended contract there). Making the UI part of the bundle in the future makes sense, in which case I don't have to worry about it any more :)

**Comment by francisco-dlp** [comment]
Great, is this ready for merging then?

**Comment by vidartf** [comment]
I haven't really tested it, but my concerns about the outline solution is at least satisfied. If that's sufficient I can merge it?

**Comment by francisco-dlp** [comment]
I think that it's find to merge it. We can all make a final test once appveyor builds the installers.

**Pull request merged**